### PR TITLE
Add Container Lifecycle support

### PR DESCRIPTION
**Pull request by vparfonov**
Signed-off-by: Vitalii Parfonov <vparfono@redhat.com>

### What does this PR do?
Add Container Lifecycle support. 
This should allow postStart and preStop actions for plugin containers. 
More information here: https://kubernetes.io/docs/tasks/configure-pod-container/attach-handler-lifecycle-event/
depends on: https://github.com/eclipse/che-plugin-broker/pull/98

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/16464

### CODE REVIEW

**Comment by dmytro-ndp** [comment]
crw-ci-test

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/model/Exec.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(command);
+  }
+
+  @Override
+  public String toString() {
+    StringBuilder sb = new StringBuilder();
```
Comment (metlos) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/model/Exec.java]: While interesting, this is not a standard `toString()` impl  in the Che codebase. We generally do not use multi-line string representation nor do we use "class" prefix. We generally do however enclose string fields in single quotes and use simple string concatenation to build the string. Please look around in the sibling classes for guidance.

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/model/Handler.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(exec);
+  }
+
+  @Override
+  public String toString() {
+    StringBuilder sb = new StringBuilder();
```
Comment (metlos) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/model/Handler.java]: `toString()` formatting.

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/model/Lifecycle.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(postStart, preStop);
+  }
+
+  @Override
+  public String toString() {
+    StringBuilder sb = new StringBuilder();
```
Comment (metlos) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/model/Lifecycle.java]: `toString()` formatting.

File: wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParserTest.java [hunk 4]
```diff
@@ -50,6 +53,16 @@ public void setUp() throws Exception {
...
   }
 
+  @Test
+  public void test() {
+    Lifecycle lifecycle = new Lifecycle();
+    lifecycle
+        .postStart(new Handler().exec(new Exec().command(Collections.singletonList("test"))))
+        .preStop(new Handler().exec(new Exec().command(Collections.singletonList("test"))));
+
+    System.out.println(lifecycle.toString());
```
Comment (metlos) [inline wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParserTest.java]: This is not testing anything.

**Review by metlos — COMMENTED** [review]
Generally LGTM apart from 2 minor objections.

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/model/Exec.java (hunk 1 above)
Comment (vparfonov) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/model/Exec.java]: I repeat how it done in classes [Volume](https://github.com/eclipse/che/blob/master/wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/model/Volume.java#L83),  [EnvVar](https://github.com/eclipse/che/blob/master/wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/model/EnvVar.java#L74), [Command](https://github.com/eclipse/che/blob/master/wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/model/Command.java#L95)

File: wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParserTest.java (hunk 4 above)
Comment (vparfonov) [reply wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParserTest.java]: yup, thanks, commit it by mistake

**Review by sleshchenko — COMMENTED** [review]
I wonder if preStop handlers are run even when we forcibly remove pods
https://github.com/eclipse/che/blob/master/assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties#L413

**Comment by vparfonov** [comment]
> I wonder if preStop handlers are run even when we forcibly remove pods
> https://github.com/eclipse/che/blob/master/assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties#L413

For normal work preStop handlers we need to increase `terminationGracePeriodSeconds` but value depends on handler itself so this open discussion, and we can increase it in separate PR.
This is note from official k8s documentations: 

_If one of the Pod’s containers has defined a preStop hook, it is invoked inside of the container. If the preStop hook is still running after the grace period expires, step 2 is then invoked with a small (2 second) one-time extended grace period. You must modify terminationGracePeriodSeconds if the preStop hook needs longer to complete._

File: wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParserTest.java (hunk 4 above)
Comment (ibuziuk) [reply wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParserTest.java]: would be great to add some coverage for the new lifecycle handlers

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/model/Exec.java (hunk 1 above)
Comment (vparfonov) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/model/Exec.java]: fixed

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/model/Lifecycle.java (hunk 3 above)
Comment (vparfonov) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/model/Lifecycle.java]: fixed

File: wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParserTest.java (hunk 4 above)
Comment (vparfonov) [reply wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParserTest.java]: fixed

**Comment by amisevsk** [comment]
> If one of the Pod’s containers has defined a preStop hook, it is invoked inside of the container. If the preStop hook is still running after the grace period expires, step 2 is then invoked with a small (2 second) one-time extended grace period. You must modify terminationGracePeriodSeconds if the preStop hook needs longer to complete.

The reason we initially set the termination grace period to 0 is that workspaces would be in a terminating state until it expired, so maybe some process in the workspace isn't properly respecting signals?

File: wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParserTest.java (hunk 4 above)
Comment (ibuziuk) [inline wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParserTest.java]: @vparfonov looks like still the test does nothing

Comment (vparfonov) [reply wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParserTest.java]: thx, fixed now

**Comment by vparfonov** [comment]
So @davidfestal @ibuziuk @amisevsk @sleshchenko @l0rd what do you think about `terminationGracePeriodSeconds`? Should we increase it in this PR?

File: wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParserTest.java [hunk 5]
```diff
@@ -28,7 +28,10 @@
 import java.util.Map;
 import org.eclipse.che.api.workspace.server.devfile.FileContentProvider;
 import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.wsplugins.model.Exec;
```
Comment (ibuziuk) [inline wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParserTest.java]: looks like still some issues here - the only change in the file is extra imports :man_shrugging:

**Comment by ibuziuk** [comment]
> what do you think about terminationGracePeriodSeconds? Should we increase it in this PR?

IMHO, we should make this property configurable via toggle per-workspace smth. similar to what is proposed in https://github.com/eclipse/che/issues/15960

File: wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParserTest.java (hunk 5 above)
Comment (vparfonov) [reply wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParserTest.java]: fixed

**Comment by vparfonov** [comment]
> > what do you think about terminationGracePeriodSeconds? Should we increase it in this PR?
> 
> IMHO, we should make this property configurable via toggle per-workspace smth. similar to what is proposed in #15960

So it this case it will be done via separate PR, right?

**Comment by ibuziuk** [comment]
> So it this case it will be done via separate PR, right?

sure

**Comment by vparfonov** [comment]
crw-ci-test

**Review by ibuziuk — APPROVED** [review]
LGTM, having some tests would be really nice to have though ;)

**Comment by vparfonov** [comment]
crw-ci-test

**Comment by vparfonov** [comment]
@metlos @sleshchenko @amisevsk can i merge it?

**Comment by ibuziuk** [comment]
@metlos @sleshchenko folks, are you ok to merge this PR?

**Review by amisevsk — APPROVED** [review]
LGTM, sorry about the delay.

I agree with Ilya that TerminationGracePeriod should be configurable (separate PR/issue), but it might open the door to some subtle configuration issues that are hard to notice. I also think there's some work to be done in implementing a nonzero termination grace period, as in my experience workspaces will take the entire termination grace period (even if it's a minute) -- that may have changed somewhere along the line though.

**Review by sleshchenko — APPROVED** [review]

**Pull request merged**